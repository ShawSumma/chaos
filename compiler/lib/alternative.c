#include "alternative.h"

void freeEverything() {
    free(main_function);
    free(scopeless);
    freeAllSymbols();
    freeAllFunctions();
    freeModulesBuffer();
    freeFunctionNamesBuffer();
    free(function_names_buffer.arr);
    if (strlen(decision_buffer) > 0) free(decision_buffer);
    freeModulePathStack();
    freeModuleStack();
    freeComplexModeStack();
    freeLeftRightBracketStack();
    freeFreeStringStack();
    freeNestedComplexModeStack();
    free(function_call_stack.arr);
}

void breakLoop() {
}

void continueLoop() {
}
