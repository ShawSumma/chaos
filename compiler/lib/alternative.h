#ifndef ALTERNATIVE_H
#define ALTERNATIVE_H

#include "../../interpreter/function.h"
enum Phase phase = INIT_PROGRAM;

void freeEverything();
void breakLoop();
void continueLoop();

char *program_file_path = "main.kaos";

#endif
