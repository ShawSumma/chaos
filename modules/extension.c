#include "extension.h"

void callRegisterInDynamicLibrary(char* dynamic_library_path) {
    kaos.defineFunction = defineFunction;
    kaos.getVariableBool = getVariableBool;
    kaos.getVariableInt = getVariableInt;
    kaos.getVariableFloat = getVariableFloat;
    kaos.getVariableString = getVariableString;
    kaos.getArrayElementBool = getArrayElementBool;
    kaos.getArrayElementInt = getArrayElementInt;
    kaos.getArrayElementFloat = getArrayElementFloat;
    kaos.getArrayElementString = getArrayElementString;
    kaos.getDictElementBool = getDictElementBool;
    kaos.getDictElementInt = getDictElementInt;
    kaos.getDictElementFloat = getDictElementFloat;
    kaos.getDictElementString = getDictElementString;
    kaos.returnVariableBool = returnVariableBool;
    kaos.returnVariableInt = returnVariableInt;
    kaos.returnVariableFloat = returnVariableFloat;
    kaos.returnVariableString = returnVariableString;
    kaos.createVariableBool = createVariableBool;
    kaos.createVariableInt = createVariableInt;
    kaos.createVariableFloat = createVariableFloat;
    kaos.createVariableString = createVariableString;
    kaos.startBuildingArray = startBuildingArray;
    kaos.returnArray = returnArray;
    kaos.startBuildingDict = startBuildingDict;
    kaos.returnDict = returnDict;
    kaos.returnComplex = returnComplex;
    lib_func func = getFunctionFromDynamicLibrary(dynamic_library_path, __KAOS_EXTENSION_REGISTER_FUNCTION__);
    func(kaos);
    if (is_interactive) {
        phase = PROGRAM;
    }
}

void callFunctionFromDynamicLibrary(_Function* function) {
    char* function_name = "";
    function_name = strcat_ext(function_name, __KAOS_EXTENSION_FUNCTION_PREFIX__);
    function_name = strcat_ext(function_name, function->name);
    lib_func func = getFunctionFromDynamicLibrary(function->module_context, function_name);
    func();
    free(function_name);
}

lib_func getFunctionFromDynamicLibrary(char* dynamic_library_path, char* function_name) {
    void     *handle  = NULL;
    lib_func  func    = NULL;

    handle = OPENLIB(dynamic_library_path);

    if (handle == NULL) {
        #if !defined(_WIN32) && !defined(_WIN64) && !defined(__CYGWIN__)
        fprintf(stderr, "Unable to open lib: %s\n", dlerror());
        #endif
    }
    func = LIBFUNC(handle, function_name);

    if (func == NULL) {
        fprintf(stderr, "Unable to get symbol\n");
    }

    return func;
}

void returnVariable(Symbol* symbol) {
    scope_override = executed_function->parent_scope;
    executed_function->symbol = createCloneFromSymbol(
        NULL,
        symbol->type,
        symbol,
        symbol->secondary_type
    );
    decision_symbol_chain = createCloneFromSymbol(
        NULL,
        symbol->type,
        symbol,
        symbol->secondary_type
    );
    scope_override = NULL;
    removeSymbol(symbol);
}
