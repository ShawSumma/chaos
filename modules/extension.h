#ifndef EXTENSION_H
#define EXTENSION_H

#include <stdio.h>
#if defined(_WIN32) || defined(_WIN64) || defined(__CYGWIN__)
#include <windows.h>
#define LIBTYPE HINSTANCE
#define OPENLIB(libname) LoadLibraryW(L"libname")
#define LIBFUNC(lib, fn) GetProcAddress((lib), (fn))
#else
#include <dlfcn.h>
#define LIBTYPE void*
#define OPENLIB(libname) dlopen((libname), RTLD_NOW | RTLD_GLOBAL)
#define LIBFUNC(lib, fn) dlsym((lib), (fn))
#endif

#include "../functions/function.h"

#if defined(_WIN32) || defined(_WIN64) || defined(__CYGWIN__)
typedef long long int (*lib_func)();
#else
typedef void (*lib_func)();
#endif

void callRegisterInSharedObject(char* so_path);
void callFunctionFromSharedObject(_Function* function);
lib_func getFunctionFromSharedObject(char* so_path, char* function_name);

#endif
