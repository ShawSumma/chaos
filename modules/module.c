#include "module.h"

void initMainContext() {
    modules_buffer.capacity = 0;
    modules_buffer.size = 0;
    module_path_stack.capacity = 0;
    module_path_stack.size = 0;
    module_stack.capacity = 0;
    module_stack.size = 0;

    char *module_path_with_extension = malloc(1 + strlen(program_file_path));
    strcpy(module_path_with_extension, program_file_path);
    pushModuleStack(module_path_with_extension, "");
    free(module_path_with_extension);
}

void appendModuleToModuleBuffer(char *name) {
    append_to_array(&modules_buffer, name);
    free(name);
}

void prependModuleToModuleBuffer(char *name) {
    prepend_to_array(&modules_buffer, name);
}

void handleModuleImport(char *module_name, bool directly_import) {
    char *module_path = "";
    char *module_dir;

    module_dir = malloc(strlen(module_path_stack.arr[module_path_stack.size - 1]) + 1);
    strcpy(module_dir, module_path_stack.arr[module_path_stack.size - 1]);
    char *ptr = strrchr(module_dir, __PATH_SEPARATOR_ASCII__);
    if (ptr) {
        *ptr = '\0';
    } else {
        module_dir = (char *) realloc(module_dir, strlen("") + 1);
        strcpy(module_dir, "");
    }

    module_path = strcat_ext(module_path, module_dir);
    if (module_path[0] != '\0') module_path = strcat_ext(module_path, __PATH_SEPARATOR__);

    for (int i = 0; i < modules_buffer.size; i++) {
        module_path = strcat_ext(module_path, modules_buffer.arr[i]);
        if (i + 1 != modules_buffer.size) {
            module_dir = (char *) realloc(module_dir, strlen(module_path) + 1);
            strcpy(module_dir, module_path);
            module_path = strcat_ext(module_path, __PATH_SEPARATOR__);
        }
    }

    char *module = modules_buffer.arr[modules_buffer.size - 1];
    if (module_name != NULL) {
        module = module_name;
    }
    if (directly_import) {
        module = "";
    }

    module_path = strcat_ext(module_path, ".");
    module_path = strcat_ext(module_path, __LANGUAGE_FILE_EXTENSION__);

#if !defined(_WIN32) && !defined(_WIN64) && !defined(__CYGWIN__)
    relative_path_to_absolute(module_path);
#else
    char actual_path[PATH_MAX];
    _fullpath(actual_path, module_path, PATH_MAX);
    module_path = (char *) realloc(module_path, strlen(actual_path) + 1);
    strcpy(module_path, actual_path);
#endif
    pushModuleStack(module_path, module);

    freeModulesBuffer();

    parseTheModuleContent(module_path);

    freeFunctionNamesBuffer();

    popModuleStack();

    free(module_path);
    free(module_dir);
    free(module_name);
}

void freeModulesBuffer() {
    for (int i = 0; i < modules_buffer.size; i++) {
        free(modules_buffer.arr[i]);
    }
    if (modules_buffer.size > 0) free(modules_buffer.arr);
    modules_buffer.capacity = 0;
    modules_buffer.size = 0;
}

void pushModuleStack(char *module_path, char *module) {
    append_to_array(&module_path_stack, module_path);
    append_to_array(&module_stack, module);
}

void popModuleStack() {
    free(module_path_stack.arr[module_path_stack.size - 1]);
    module_path_stack.size--;
    free(module_stack.arr[module_stack.size - 1]);
    module_stack.size--;
}

void freeModulePathStack() {
    for (int i = 0; i < module_path_stack.size; i++) {
        free(module_path_stack.arr[i]);
    }
    module_path_stack.capacity = 0;
    module_path_stack.size = 0;
    free(module_path_stack.arr);
}

void freeModuleStack() {
    for (int i = 0; i < module_stack.size; i++) {
        free(module_stack.arr[i]);
    }
    module_stack.capacity = 0;
    module_stack.size = 0;
    free(module_stack.arr);
}

char* getCurrentModule() {
    return module_path_stack.arr[module_path_stack.size - 1];
}
