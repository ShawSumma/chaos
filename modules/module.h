#ifndef MODULE_H
#define MODULE_H

#include <stdio.h>

#include "../symbol.h"
#include "../errors.h"
#include "../utilities/injector.h"

string_array modules_buffer;

extern char *program_file_path;
extern char *program_file_dir;
string_array module_path_stack;
string_array module_stack;

extern int module_parsing;

void initMainContext();
void appendModuleToModuleBuffer(char *name);
void prependModuleToModuleBuffer(char *name);
void handleModuleImport(char *module_name, bool directly_import);
void freeModulesBuffer();
void pushModuleStack(char *module_path, char *module);
void popModuleStack();
void freeModulePathStack();
void freeModuleStack();
char* getCurrentModule();

#endif
