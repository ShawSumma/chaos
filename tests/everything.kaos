/*
This is a test that almost every feature of the language
used in a program altogether to see if anything is wrong.
*/

import modules.lib
from modules.lib import everything_f2
import modules.lib as lib2

/*
This is a
multiline comment
for the function below
@description Adds to numbers together
*/
num def add(num x, num y)
    num z = x + y

    bool a = true
    num b = 576.34
    // This is a comment
    list c = ['foo', "bar", 17, 3.14, a, false] // End of Line(EOL) comment
    pretty echo c
    # This is a comment too
    dict d = {'a': 1, 'b': b, 'c': [2, 3, 4]} # End of Line(EOL) comment
    pretty print d

    del c
end {
    z == 8  : f1(),
    z > 10  : f2(),
    default : f3()
}

print add(3, 5)

echo lib.everything_f1(4)
print everything_f2(32)
print lib2.everything_f3(121, 34)

num def f1()
    num a = 101
    return a
end

num def f2()
    num b = 102
    return b
end

num def f3()
    num c = 103
    return c
end

list a = [
    1,
    2,
    3
]

foreach a as el
    print el
    el = 5
    2 times do
        print el
    end
end

dict n = {'a': 'foo', 'b': 'bar', 'c': 'baz'}

foreach n as key : val
    print key
    print val
end
