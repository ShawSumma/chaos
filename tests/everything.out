[
    'foo',
    'bar',
    17,
    3.14,
    true,
    false
]{
    'a': 1,
    'b': 576.34,
    'c': [
        2,
        3,
        4
    ]
}
101
264
465
1
5
5
2
5
5
3
5
5
a
foo
b
bar
c
baz
